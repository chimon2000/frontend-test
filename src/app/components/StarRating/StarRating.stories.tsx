import * as React from "react";
import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import { action } from "@storybook/addon-actions";
import { withKnobs, select } from "@storybook/addon-knobs";

import { StarRating } from "./StarRating";

const stories = storiesOf("Components", module);
stories.add(
  "StarRating",
  withInfo({ inline: true })(() => (
    <StarRating
      size={select("Size", ["small", "large"], "small")}
      rating={select("Rating", [1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5], 2.5)}
    />
  ))
);
