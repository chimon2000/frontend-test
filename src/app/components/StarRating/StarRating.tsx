import * as React from "react";
import { times } from "lodash";

import Icon from "../Icon";

export type RatingRange = 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5;

export interface StarRatingProps {
  className?: string;
  size?: "small" | "large";
  rating: RatingRange;
}

export const TYPE_MAP = {
  whole: "star",
  half: "star_half",
  empty: "star_border"
};

const getStars = (count: number, type: "whole" | "half" | "empty") => {
  const result = [];
  times(count, (n: number) => {
    result.push(
      <Icon className="star-rating__star" key={`${type}-${n}`}>
        {TYPE_MAP[type]}
      </Icon>
    );
  });

  return result;
};

export const StarRating: React.FunctionComponent<StarRatingProps> = ({
  children,
  className,
  size = "small",
  rating
}) => {
  const whole = Math.floor(rating);
  const half = rating > whole ? 1 : 0;
  const empty = 5 - Math.ceil(rating);
  return (
    <div className={`star-rating star-rating--${size}`}>
      {getStars(whole, "whole")}
      {getStars(half, "half")}
      {getStars(empty, "empty")}
    </div>
  );
};
