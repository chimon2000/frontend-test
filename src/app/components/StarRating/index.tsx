export {
  StarRating as default,
  RatingRange,
  StarRatingProps
} from "./StarRating";
