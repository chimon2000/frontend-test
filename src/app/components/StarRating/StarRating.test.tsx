import * as React from "react";
import { shallow } from "enzyme";

import { StarRating, TYPE_MAP } from "./StarRating";

describe("<StarRating />", () => {
  it("doesn't blow up", () => {
    const wrapper = shallow(<StarRating rating={4} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("renders 5 whole stars", () => {
    const wholeStarText = TYPE_MAP["whole"];
    const wrapper = shallow(<StarRating rating={5} />);

    const texts = wrapper
      .find(".star-rating__star")
      .map(node => node.children().text());
    expect(texts).toEqual([
      wholeStarText,
      wholeStarText,
      wholeStarText,
      wholeStarText,
      wholeStarText
    ]);
  });

  it("renders 2.5 of 5 stars", () => {
    const wrapper = shallow(<StarRating rating={2.5} />);

    const texts = wrapper
      .find(".star-rating__star")
      .map(node => node.children().text());
    expect(texts).toEqual([
      TYPE_MAP["whole"],
      TYPE_MAP["whole"],
      TYPE_MAP["half"],
      TYPE_MAP["empty"],
      TYPE_MAP["empty"]
    ]);
  });
});
