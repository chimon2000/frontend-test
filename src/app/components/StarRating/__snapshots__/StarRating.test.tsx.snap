// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StarRating /> doesn't blow up 1`] = `
<div
  className="star-rating star-rating--small"
>
  <Component
    className="star-rating__star"
    key="whole-0"
  >
    star
  </Component>
  <Component
    className="star-rating__star"
    key="whole-1"
  >
    star
  </Component>
  <Component
    className="star-rating__star"
    key="whole-2"
  >
    star
  </Component>
  <Component
    className="star-rating__star"
    key="whole-3"
  >
    star
  </Component>
  <Component
    className="star-rating__star"
    key="empty-0"
  >
    star_border
  </Component>
</div>
`;
